design/baremetal-operator/unmanaged-annotation.md

```diff
@@ -0,0 +1,106 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Unmanaged annotation
+
+## Status
+
+implementable
+
+## Summary
+
+Provide a way to prevent management of BaremetalHost resources
+after provisioning is completed, to facilitate the pivot of
+BaremetalHost resources in a multi-tier deployment.
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+There is an existing
+[pause annotation](https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md#pausing-reconciliation)
+which pauses reconciliation of the BMH resources, but this does not remove
+the underlying Ironic host, so power management actions may be performed
+even when the BMH is marked as `paused`.
+
+### Goals
+
+* Add an API to disable management of a BMH on the parent cluster, including
+  all power management actions
+* Make it possible to delete a BMH resource in this "unmanaged" without
+  triggering deprovisioning
+* Ensure it is possible to restart management of BMH resources (in the case
+  where they are not deleted from the parent cluster)
+* Avoid changing behavior of the existing `paused` annotation since
+
+### Non-Goals
+
+* Any coordination between the tiers of BMH resources, that has to be handled externally
+
+## Proposal
+
+### Expected workflow
+
+* User creates BMH resource(s) in parent cluster
+* Provisioning of BMH triggered, which results in a running child cluster
+* Parent cluster BMH resources marked as unmanaged
+* Child cluster BMH resources created, with BMC credentials, but marked
+  externallyProvisioned: true
+
+At this point, the physical hosts are owned by the child cluster BMH, but the
+inventory still exists in the parent cluster.
+
+In the event that all of the child cluster hosts require reprovisioning, it
+would be necessary to remove the maintenance/unmanaged flag, or
+delete/re-create the BMH resource so that the process starts again at the
+beginning.
+
+### API changes
+
+Add support for a new annotation, where the key is `baremetalhost.metal3.io/unmanaged`
+
+```yaml
+baremetalhost.metal3.io/unmanaged: ""
+```
+
+The value is ignored, similar to the `paused` annotation and could optionally
+include context from the system/user which applies the annotation.
+
+This annotation will be evaluated early in the `Reconcile()` loop, but after the
+`paused` `status` and `hardwaredetails` annotations are evaluated.
```

Comment:
Why do we want to do it after those other annotations are handled? Does it actually matter, or is that just what we're doing?

I think we want to do it after the hardware details and status annotations are handled to avoid triggering inspection when the unmanaged annotation is set?

It seems like want to handle it before the paused annotation, since unmanaged is a more comprehensive way to say "stop paying attention to this host" than paused is.

Would it be safer for this new annotation to only apply to hosts in a steady state? What happens if provisioning starts and this annotation is added, for example? Or deprovisioning?

Reply:
It definitely must be handled after paused and status. I think in the case of hardwaredetails that might just be where we're planning on doing it rather than a strict requirement.

The question of what happens in a non-steady state is a good one. This might be a good reason to use the Unmanaged provisioning state as we discussed, since it means the host will always return to the Registering or Provisoned states afterwards. We might want to ignore the annotation if we are in the Deprovisioning state though, because AFAIR there's no way to distinguish whether the host was Provisioned or Deprovisioning before going to Unmanaged.

Reply:
>  I think in the case of hardwaredetails that might just be where we're planning on doing it rather than a strict requirement

Yes it's not a strict requirement, but I was thinking there's not really any reason to prevent consuming hardwaredetails while in the unmanaged mode, thus what I proposed in https://github.com/metal3-io/baremetal-operator/pull/827

> The question of what happens in a non-steady state is a good one. This might be a good reason to use the Unmanaged provisioning state as we discussed, since it means the host will always return to the Registering or Provisoned states afterwards

Agreed I'd not really considered this case - previous feedback from @dhellmann was that we wouldn't want to add a new transition to the `Unmanaged` state, but it seems like the cleanest way to enforce that is to handle this mode via the state machine?

The other option is to only allow this annotation to function when in a specific steady state (e.g for the pivot use-case, only `Provisioned` would be enough)

Reply:
> It definitely must be handled after paused and status. I think in the case of hardwaredetails that might just be where we're planning on doing it rather than a strict requirement.

If we handle this after the paused annotation, that means pausing has higher precedence than marking it as unmanaged. I'm not sure that makes sense.


> > The question of what happens in a non-steady state is a good one. This might be a good reason to use the Unmanaged provisioning state as we discussed, since it means the host will always return to the Registering or Provisoned states afterwards
> 
> Agreed I'd not really considered this case - previous feedback from @dhellmann was that we wouldn't want to add a new transition to the `Unmanaged` state, but it seems like the cleanest way to enforce that is to handle this mode via the state machine?
> 
> The other option is to only allow this annotation to function when in a specific steady state (e.g for the pivot use-case, only `Provisioned` would be enough)

Can we handle it in `actionManageSteadyState` in the controller?

Reply:
> If we handle this after the paused annotation, that means pausing has higher precedence than marking it as unmanaged. I'm not sure that makes sense.

Pausing is meant to appear to stop reconciliation, so the only way it _does_ make sense is if it has higher precedence than everything.

> Can we handle it in `actionManageSteadyState` in the controller?

No, because it would get re-registered first.

Reply:
Yes, it is really a requirement for pivoting as done with CAPI and CAPM3, that the pausing annotation is the first thing handled and has the highest precendence